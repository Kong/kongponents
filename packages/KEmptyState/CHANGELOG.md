# @kongponents/kcard

## Table of Contents

- [0.1.3](#013)

### 0.1.3
 - add `body` only support [#108](https://github.com/Kong/kongponents/pull/108)
