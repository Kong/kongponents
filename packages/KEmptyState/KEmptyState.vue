<template>
  <div class="empty-state-wrapper">
    <div class="empty-state-title">
      <h5><slot name="title"/></h5>
    </div>
    <div class="empty-state-content">
      <p><slot name="message"/></p>
      <p>
        <slot name="cta">
          <KButton
            v-if="!ctaIsHidden"
            :is-rounded="true"
            appearance="outline-primary"
            @click.native="handleClick">
            {{ ctaText }}
          </KButton>
        </slot>
      </p>
    </div>
  </div>
</template>

<script>
import KButton from '@kongponents/kbutton'
export default {
  name: 'KEmptyState',
  components: { KButton },
  props: {
    ctaIsHidden: {
      type: Boolean,
      default: false
    },
    ctaText: {
      type: String,
      default: ''
    },
    handleClick: {
      type: Function,
      default: () => () => null
    }
  }
}
</script>

<style scoped>
  .empty-state-wrapper {
  padding: 4rem 0;
  text-align: center;
  color: rgba(0,0,0,.70);
  border: 1px solid rgba(151,151,151,.1);
  border-radius: 3px;
  background-color: rgba(0,0,0,.03);
}
.empty-state-wrapper h5 {
  margin: 0 0 .75rem;
  font-size: 18px;
  font-weight: 500;
  line-height: 21px;
}
.empty-state-wrapper p {
  margin: 0 auto 1.5rem;
  font-size: 1rem;
  line-height: 1.375rem;
  font-weight: 400;
  max-width: 50%;
}
.empty-state-wrapper p:last-child {
  margin: 0;
  margin-left: auto;
  margin-right: auto;
}
</style>
