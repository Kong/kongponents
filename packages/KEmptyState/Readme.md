```js
<KEmptyState 
  cta-text="Do A Thing">
  <template slot="title">No Enabled Plugins</template>
  <template slot="message">To enable, click below.</template> 
</KEmptyState>
```
