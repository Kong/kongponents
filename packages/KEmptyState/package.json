{
  "name": "@kongponents/kemptystate",
  "version": "0.1.7",
  "description": "Empty State component",
  "main": "KEmptyState.vue",
  "scripts": {
    "test": "echo \"Error: no test specified\""
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/Kong/kongponents.git"
  },
  "author": "Kong Inc",
  "license": "ISC",
  "bugs": {
    "url": "https://github.com/Kong/kongponents/issues"
  },
  "homepage": "https://github.com/Kong/kongponents#readme",
  "dependencies": {
    "@kongponents/kbutton": "^0.1.7"
  }
}
