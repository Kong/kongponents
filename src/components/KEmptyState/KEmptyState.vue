<template>
  <div
    class="k-empty-state"
    :class="[iconVariant]"
  >
    <div class="empty-state-content">
      <div
        :aria-hidden="$slots.image ? undefined : 'true'"
        :class="iconContainerClass"
      >
        <slot name="image">
          <slot name="icon">
            <component
              :is="getEmptyStateIcon"
              :color="getIconColor"
              decorative
              :size="KUI_ICON_SIZE_60"
            />
          </slot>
        </slot>
      </div>
      <div
        v-if="title || $slots.title"
        class="empty-state-title"
      >
        <slot name="title">
          {{ title }}
        </slot>
      </div>
      <div
        v-if="message || $slots.default"
        class="empty-state-message"
      >
        <slot name="default">
          <p>
            {{ message }}
          </p>
        </slot>
      </div>
      <div
        v-if="(actionButtonVisible && actionButtonText) || $slots.action"
        class="empty-state-action"
      >
        <slot name="action">
          <KButton
            :disabled="actionButtonDisabled"
            type="button"
            @click="emit('click-action')"
          >
            <slot name="action-button-icon" />

            {{ actionButtonText }}
          </KButton>
        </slot>
      </div>
    </div>
    <div
      v-if="features.length"
      class="empty-state-features-container"
    >
      <template
        v-for="(feature, idx) in features"
        :key="`feature-${componentId}-${feature.key || idx}`"
      >
        <KCard class="empty-state-feature-card">
          <template #title>
            <div
              v-if="feature.key && $slots[`feature-icon-${feature.key}`]"
              aria-hidden="true"
              class="feature-icon"
            >
              <slot :name="`feature-icon-${feature.key}`" />
            </div>

            <div class="card-header">
              {{ feature.title }}
            </div>
          </template>

          <div :title="feature.description">
            {{ feature.description }}
          </div>
        </KCard>
      </template>
    </div>
    <div
      v-if="$slots.footer"
      class="empty-state-footer-container"
    >
      <slot name="footer" />
    </div>
  </div>
</template>

<script lang="ts" setup>
import { computed, useId } from 'vue'
import { AnalyticsIcon, WarningIcon, SearchIcon, KongIcon } from '@kong/icons'
import { KUI_COLOR_TEXT_DECORATIVE_AQUA, KUI_COLOR_TEXT_NEUTRAL, KUI_COLOR_TEXT_WARNING, KUI_ICON_SIZE_60 } from '@kong/design-tokens'
import KButton from '@/components/KButton/KButton.vue'
import { EmptyStateIconVariants } from '@/types'
import type { EmptyStateProps, EmptyStateEmits, EmptyStateSlots } from '@/types'
import KCard from '@/components/KCard/KCard.vue'

type EmptyStateIcon = typeof AnalyticsIcon // all icons are the same type so we can use any of them

const {
  title = '',
  message = '',
  actionButtonText = '',
  actionButtonVisible = true,
  actionButtonDisabled,
  iconVariant = EmptyStateIconVariants.Default,
  iconBackground = false,
  features = [],
} = defineProps<EmptyStateProps>()

const emit = defineEmits<EmptyStateEmits>()

const slots = defineSlots<EmptyStateSlots>()

const componentId = useId()

const getEmptyStateIcon = computed((): EmptyStateIcon => {
  switch (iconVariant) {
    case EmptyStateIconVariants.Default:
      return AnalyticsIcon
    case EmptyStateIconVariants.Error:
      return WarningIcon
    case EmptyStateIconVariants.Search:
      return SearchIcon
    case EmptyStateIconVariants.Kong:
      return KongIcon
    default:
      return AnalyticsIcon // default variant in case of invalid value
  }
})

const getIconColor = computed((): string => {
  if (iconBackground) {
    return KUI_COLOR_TEXT_DECORATIVE_AQUA
  }

  switch (iconVariant) {
    case EmptyStateIconVariants.Error:
      return KUI_COLOR_TEXT_WARNING
    default:
      return KUI_COLOR_TEXT_NEUTRAL
  }
})

const iconContainerClass = computed((): string => {
  if (slots.image) {
    return 'empty-state-image'
  }

  const defaultIconClass = 'empty-state-icon'
  if (iconBackground) {
    return `${defaultIconClass} has-background`
  }

  return defaultIconClass
})
</script>

<style lang="scss" scoped>
/* Component variables */

$kEmptyStateMaxWidth: 640px;

/* Component styles */

.k-empty-state {
  align-items: center;
  background-color: var(--kui-color-background, $kui-color-background);
  box-sizing: border-box;
  display: flex;
  flex-direction: column;
  font-family: var(--kui-font-family-text, $kui-font-family-text);
  gap: var(--kui-space-90, $kui-space-90);
  padding: var(--kui-space-90, $kui-space-90);
  width: 100%;

  @media (min-width: $kui-breakpoint-phablet) {
    padding: var(--kui-space-110, $kui-space-110);
  }

  .empty-state-content {
    align-items: center;
    display: flex;
    flex-direction: column;
    gap: var(--kui-space-70, $kui-space-70);
    text-align: center;
    width: 100%;

    .empty-state-icon {
      color: var(--kui-color-text-neutral, $kui-color-text-neutral);

      :deep(#{$kongponentsKongIconSelector}) {
        height: var(--kui-icon-size-60, $kui-icon-size-60) !important;
        width: var(--kui-icon-size-60, $kui-icon-size-60) !important;
      }

      &.has-background {
        background-color: var(--kui-color-background-decorative-aqua-weakest, $kui-color-background-decorative-aqua-weakest);
        border-radius: var(--kui-border-radius-20, $kui-border-radius-20);
        color: var(--kui-color-text-decorative-aqua, $kui-color-text-decorative-aqua);
        padding: var(--kui-space-40, $kui-space-40);

        :deep(#{$kongponentsKongIconSelector}) {
          height: var(--kui-icon-size-50, $kui-icon-size-50) !important;
          width: var(--kui-icon-size-50, $kui-icon-size-50) !important;
        }
      }
    }

    .empty-state-image {
      max-width: $kEmptyStateMaxWidth;
      overflow: hidden;
      width: 100%;
    }

    .empty-state-title {
      color: var(--kui-color-text, $kui-color-text);
      font-size: var(--kui-font-size-50, $kui-font-size-50);
      font-weight: var(--kui-font-weight-bold, $kui-font-weight-bold);
      line-height: var(--kui-line-height-40, $kui-line-height-40);
      max-width: 570px; // limit width so the text stays readable if title is too long
    }

    .empty-state-message {
      color: var(--kui-color-text-neutral, $kui-color-text-neutral);
      display: flex;
      flex-direction: column;
      font-size: var(--kui-font-size-30, $kui-font-size-30);
      font-weight: var(--kui-font-weight-regular, $kui-font-weight-regular);
      gap: var(--kui-space-70, $kui-space-70);
      line-height: var(--kui-line-height-30, $kui-line-height-30);
      max-width: $kEmptyStateMaxWidth; // limit width so the message stays readable if title is too long

      p {
        margin: var(--kui-space-0, $kui-space-0);
      }
    }

    .empty-state-action {
      align-items: center;
      display: flex;
      gap: var(--kui-space-40, $kui-space-40);
    }
  }

  .empty-state-features-container {
    display: grid;
    gap: var(--kui-space-70, $kui-space-70);
    grid-template-columns: 1fr;
    justify-content: space-around;
    max-width: $kEmptyStateMaxWidth;

    @media (min-width: $kui-breakpoint-phablet) {
      grid-template-columns: 1fr 1fr;
    }

    .empty-state-feature-card {
      background-color: var(--kui-color-background-neutral-weakest, $kui-color-background-neutral-weakest);
      gap: var(--kui-space-30, $kui-space-30);

      .feature-icon {
        color: var(--kui-color-text-neutral-stronger, $kui-color-text-neutral-stronger);
        display: flex;
        margin-bottom: var(--kui-space-50, $kui-space-50);

        :deep(#{$kongponentsKongIconSelector}) {
          height: var(--kui-icon-size-40, $kui-icon-size-40) !important;
          width: var(--kui-icon-size-40, $kui-icon-size-40) !important;
        }
      }

      :deep(.card-title) {
        @include truncate(2);

        font-size: var(--kui-font-size-30, $kui-font-size-30);
        font-weight: var(--kui-font-weight-semibold, $kui-font-weight-semibold);
        line-height: var(--kui-line-height-30, $kui-line-height-30);
      }

      :deep(.card-content) {
        @include truncate(3);

        color: var(--kui-color-text-neutral, $kui-color-text-neutral);
      }
    }
  }

  .empty-state-footer-container {
    border-top: var(--kui-border-width-10, $kui-border-width-10) solid var(--kui-color-border, $kui-color-border);
    color: var(--kui-color-text-neutral, $kui-color-text-neutral);
    display: flex;
    flex-direction: column;
    font-size: var(--kui-font-size-30, $kui-font-size-30);
    font-weight: var(--kui-font-weight-regular, $kui-font-weight-regular);
    gap: $kui-space-40;
    line-height: var(--kui-line-height-30, $kui-line-height-30);
    max-width: $kEmptyStateMaxWidth; // limit width so the message stays readable if title is too long
    padding-top: $kui-space-90;
    width: 100%;
  }
}
</style>
