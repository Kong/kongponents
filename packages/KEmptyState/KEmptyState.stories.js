import { storiesOf } from '../../storybook/node_modules/@storybook/vue'
import KEmptyState from './KEmptyState.vue'

storiesOf('EmptyState', module)
  .add('Call to action', () => ({
    components: { KEmptyState },
    template: `<div>
    <KEmptyState cta-text="Do A Thing" :handle-click="callToAction">
      <template slot="title">No Enabled Entities</template>
      <template slot="message">To do the thing, click below.</template>
      </KEmptyState>
    </div>`,
    methods: {
      callToAction () {
        alert('action thing!')
      }
    }
  }), {info: {}})
