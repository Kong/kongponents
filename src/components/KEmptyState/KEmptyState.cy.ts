import { mount } from 'cypress/vue'
import KEmptyState from '@/components/KEmptyState/KEmptyState.vue'

describe('KEmptyState', () => {
  it('renders all elements correctly', () => {
    mount(KEmptyState)

    cy.get('.k-empty-state').should('be.visible')
    cy.get('.empty-state-icon').should('be.visible')
    cy.get('.empty-state-title').should('not.exist')
    cy.get('.empty-state-message').should('not.exist')
    cy.get('.empty-state-action').should('not.exist') // because action button text wasn't provided
  })

  it('renders title and message when provided', () => {
    const title = 'Title'
    const message = 'Message'

    mount(KEmptyState, {
      props: {
        title,
        message,
      },
    })

    cy.get('.empty-state-title').should('be.visible').should('contain', title)
    cy.get('.empty-state-message').should('be.visible').should('contain', message)
  })

  it('renders action button when provided button text', () => {
    const actionButtonText = 'Action'

    mount(KEmptyState, {
      props: {
        actionButtonText,
      },
    })

    cy.get('.empty-state-action').should('be.visible').should('contain', actionButtonText)
  })

  it('does not render action button when hidden', () => {
    mount(KEmptyState, {
      props: {
        actionButtonVisible: false,
        actionButtonText: 'Action',
      },
    })

    cy.get('.empty-state-action').should('not.exist')
  })

  it('correctly handles action button disabled state', () => {
    mount(KEmptyState, {
      props: {
        actionButtonText: 'Action',
        actionButtonDisabled: true,
      },
    })

    cy.get('.empty-state-action').should('be.visible').find('button').should('be.disabled')
  })

  it('displays content passed through default slot correctly', () => {
    const content = 'Content'

    mount(KEmptyState, {
      props: {
        message: 'Message',
      },
      slots: {
        default: `<span data-testid="slotted-message">${content}</span>`,
      },
    })

    cy.get('.empty-state-message').findTestId('slotted-message').should('be.visible').should('contain', content)
  })

  it('displays icon passed through icon slot', () => {
    mount(KEmptyState, {
      slots: {
        icon: '<img data-testid="slotted-icon" src="https://via.placeholder.com/36" />',
      },
    })

    cy.get('.empty-state-icon').findTestId('slotted-icon').should('be.visible')
  })

  it('emits event when action button is clicked', () => {
    mount(KEmptyState, {
      props: {
        actionButtonText: 'Action',
      },
    })

    cy.get('.empty-state-action').should('be.visible').find('button').click().then(() => {
      cy.wrap(Cypress.vueWrapper.emitted()).should('have.property', 'action-click')
    })
  })
})
