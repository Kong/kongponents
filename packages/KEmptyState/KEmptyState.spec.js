import { mount } from '@vue/test-utils'
import KEmptyState from '@/KEmptyState/KEmptyState'

describe('KEmptyState', () => {
  it('renders slots when passed', () => {
    const emptyTitle = 'No Entities Yet'
    const emptyMessage = 'To Add an Entity, Press the Button'
    const wrapper = mount(KEmptyState, {
      slots: {
        'title': `<span>${emptyTitle}</span>`,
        'message': `<div>${emptyMessage}</div>`
      }
    })

    expect(wrapper.find('.empty-state-title').html()).toEqual(expect.stringContaining(emptyTitle))
    expect(wrapper.find('.empty-state-content').html()).toEqual(expect.stringContaining(emptyMessage))
  })

  it('remains empty when no slots are passed', () => {
    const emptyTitle = ''
    const emptyMessage = ''
    const wrapper = mount(KEmptyState, {
      slots: {}
    })

    expect(wrapper.find('.empty-state-title').html()).toEqual(expect.stringContaining(emptyTitle))
    expect(wrapper.find('.empty-state-content').html()).toEqual(expect.stringContaining(emptyMessage))
  })
})
