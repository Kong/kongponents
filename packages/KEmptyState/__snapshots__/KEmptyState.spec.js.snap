// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KEmptyState matches snapshot 1`] = `
<div class="empty-state-wrapper">
  <div class="empty-state-title">
    <!---->
    <h5></h5>
  </div>
  <div class="empty-state-content">
    <p></p>
    <p><button type="button" class="k-button medium rounded outline">

      </button></p>
  </div>
</div>
`;
